code/aes/Hacl.Impl.AES.Generic.fst

```diff
@@ -0,0 +1,1800 @@
+  let h0 = ST.get() in
+  assert (v ((nr a +! 1ul) *. klen m) == v (nr a +! 1ul) * v (klen m));
+  assert (v (nr a +! 1ul) == v (nr a) + 1);
+  assert (v (nr a) == Spec.num_rounds a);
+  LSeq.eq_intro (as_seq h0 kex)
+    (Seq.create ((Spec.num_rounds a+1) * v (klen m)) (elem_zero m));
+  keyx_zero_lemma m a (as_seq h0 kex);
+  load_nonce #m n nonce;
+  let h1 = ST.get() in
+  u8_16_to_le_lemma (nonce_to_bytes m (as_seq h1 n));
+  key_expansion #m #a kex key
+
+(* BEGIN STANDARD PATTERN FOR AVOIDING INLINING *)
+inline_for_extraction noextract
+val aes128_ni_init:
+    ctx: aes_ctx MAES Spec.AES128
+  -> key: skey Spec.AES128
+  -> nonce: lbuffer uint8 12ul ->
+  Stack unit
+  (requires (fun h -> live h ctx /\ live h nonce /\ live h key /\ disjoint ctx key /\
+    as_seq h ctx == Seq.create (size_v (ctxlen MAES Spec.AES128)) (elem_zero MAES)))
+  (ensures (fun h0 b h1 -> modifies (loc ctx) h0 h1 /\
+    (let n = LSeq.sub (as_seq h1 ctx) 0 (v (nlen MAES)) in
+    let kex = LSeq.sub (as_seq h1 ctx)
+      (v (nlen MAES)) ((Spec.num_rounds Spec.AES128+1) * v (klen MAES)) in
+    let state = Spec.aes_ctr32_init_LE Spec.AES128
+      (as_seq h0 key) (as_seq h0 nonce) in
+    nonce_to_bytes MAES n == state.block /\
+    keyx_to_bytes MAES Spec.AES128 kex == state.key_ex)))
+
+let aes128_ni_init ctx key nonce = aes_init_ #MAES #Spec.AES128 ctx key nonce
+
+inline_for_extraction noextract
+val aes256_ni_init:
+    ctx: aes_ctx MAES Spec.AES256
+  -> key: skey Spec.AES256
+  -> nonce: lbuffer uint8 12ul ->
+  Stack unit
+  (requires (fun h -> live h ctx /\ live h nonce /\ live h key /\ disjoint ctx key /\
+    as_seq h ctx == Seq.create (size_v (ctxlen MAES Spec.AES256)) (elem_zero MAES)))
+  (ensures (fun h0 b h1 -> modifies (loc ctx) h0 h1 /\
+    (let n = LSeq.sub (as_seq h1 ctx) 0 (v (nlen MAES)) in
+    let kex = LSeq.sub (as_seq h1 ctx)
+      (v (nlen MAES)) ((Spec.num_rounds Spec.AES256+1) * v (klen MAES)) in
+    let state = Spec.aes_ctr32_init_LE Spec.AES256
+      (as_seq h0 key) (as_seq h0 nonce) in
+    nonce_to_bytes MAES n == state.block /\
+    keyx_to_bytes MAES Spec.AES256 kex == state.key_ex)))
+
+let aes256_ni_init ctx key nonce = aes_init_ #MAES #Spec.AES256 ctx key nonce
+
+inline_for_extraction noextract
+val aes_init:
+    #m: m_spec
+  -> #a: Spec.variant
+  -> ctx: aes_ctx m a
+  -> key: skey a
+  -> nonce: lbuffer uint8 12ul ->
+  Stack unit
+  (requires (fun h -> live h ctx /\ live h nonce /\ live h key /\
+    disjoint ctx key /\
+    as_seq h ctx == Seq.create (size_v (ctxlen m a)) (elem_zero m)))
+  (ensures (fun h0 b h1 -> modifies (loc ctx) h0 h1 /\
+    (let n = LSeq.sub (as_seq h1 ctx) 0 (v (nlen m)) in
+    let kex = LSeq.sub (as_seq h1 ctx)
+      (v (nlen m)) ((Spec.num_rounds a+1) * v (klen m)) in
+    let state = Spec.aes_ctr32_init_LE a (as_seq h0 key)
+      (as_seq h0 nonce) in
+    nonce_to_bytes m n == state.block /\
+    keyx_to_bytes m a kex == state.key_ex)))
+
+let aes_init #m #a ctx key nonce =
```

Comment:
I take it that a toolchain supports either building both variants, or none at all?